src/back/wgsl/writer.rs

```diff
@@ -1128,8 +1222,70 @@ impl<W: Write> Writer<W> {
                     self.out.write_char(COMPONENTS[sc as usize])?;
                 }
             }
-            _ => {
-                return Err(Error::Unimplemented(format!("write_expr {:?}", expression)));
+            Expression::Unary { op, expr } => {
+                let unary = match op {
+                    crate::UnaryOperator::Negate => "-",
+                    crate::UnaryOperator::Not => {
+                        match *func_ctx.info[expr].ty.inner_with(&module.types) {
+                            TypeInner::Scalar {
+                                kind: ScalarKind::Bool,
+                                ..
+                            }
+                            | TypeInner::Vector { .. } => "!",
+                            _ => "~",
+                        }
+                    }
+                };
+
+                write!(self.out, "{}(", unary)?;
+                self.write_expr(module, expr, func_ctx)?;
+
+                write!(self.out, ")")?
+            }
+            Expression::Select {
+                condition,
+                accept,
+                reject,
+            } => {
+                write!(self.out, "select(")?;
+                self.write_expr(module, accept, func_ctx)?;
+                write!(self.out, ", ")?;
+                self.write_expr(module, reject, func_ctx)?;
+                write!(self.out, ", ")?;
+                self.write_expr(module, condition, func_ctx)?;
+                write!(self.out, ")")?
+            }
+            Expression::Derivative { axis, expr } => {
+                write!(
+                    self.out,
+                    "{}(",
+                    match axis {
```

Comment:
nit: let's do minimal work within macro invocations. Macros are always opaque to IDEs, so we should move the logic out. Here, we can match outside of the macro.